### PR TITLE
Output vehicle id when splitting by vehicle

### PR DESCRIPTION
**Pull request by fonsecadeline**
Step 1 for : https://gitlab.com/mapotempo/optimizer-api/-/issues/660

### CODE REVIEW

File: test/lib/interpreters/split_clustering_functions_test.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Copyright © Mapotempo, 2021
```
Comment (senhalil) [inline test/lib/interpreters/split_clustering_functions_test.rb:1]: I don't think this PR  is the right place for needs this new file and moving `test_if_duration_from_and_to_depot_is_filled_correctly`, as we have an on-going discussion in another PR exactly for this purpose.

I think we should continue with the current rule until we finally decide what to do and after then we can do such moves in a PR dedicated to this task.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Added
+
+- Corresponding vehicle_id provided for each service when using partition with vehicle entity.
```
Comment (senhalil) [inline CHANGELOG.md:7]:
```suggestion
- Corresponding vehicle_id is returned for each service if problem is partitioned with vehicle entity [#110](https://github.com/Mapotempo/optimizer-api/pull/110)
```

**Review by senhalil — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Added
+
+- Corresponding vehicle_id is returned for each service if problem is partitioned with vehicle entity [#110](https://github.com/Mapotempo/optimizer-api/pull/110)
```
Comment (braktar) [inline CHANGELOG.md:7]:
```suggestion
- `vehicle_id` to services `skills` when using partition with vehicle entity [#110](https://github.com/Mapotempo/optimizer-api/pull/110)
```

File: test/lib/interpreters/split_clustering_test.rb [hunk 4]
```diff
@@ -126,7 +126,7 @@ def test_if_duration_from_and_to_depot_is_filled_correctly
       problem[:points].each{ |p| p.delete(:matrix_index) }
       problem[:vehicles][0].delete(:matrix_id)
       problem[:vehicles] << problem[:vehicles].first.dup
-
+
```
Comment (senhalil) [inline test/lib/interpreters/split_clustering_test.rb:129]:
```suggestion

```

File: test/lib/interpreters/split_clustering_test.rb [hunk 5]
```diff
@@ -136,7 +136,7 @@ def test_if_duration_from_and_to_depot_is_filled_correctly
         Interpreters::SplitClustering.split_balanced_kmeans({ vrp: TestHelper.create(problem), service: :demo }, problem[:vehicles].size, cut_symbol: :duration, entity: :vehicle, restarts: 1)
       end
       mock.verify # check if it is called
-
+
```
Comment (senhalil) [inline test/lib/interpreters/split_clustering_test.rb:139]:
```suggestion

```

File: test/lib/interpreters/split_clustering_test.rb (hunk 4 above)
Comment (braktar) [inline test/lib/interpreters/split_clustering_test.rb:129]: remove this extra space

File: test/lib/interpreters/split_clustering_test.rb (hunk 5 above)
Comment (braktar) [inline test/lib/interpreters/split_clustering_test.rb:139]: remove this extra space

**Review by senhalil — APPROVED** [review]
white-space typos otherwise :ok_hand:

File: test/api/v01/output_test.rb [hunk 6]
```diff
@@ -72,20 +72,32 @@ def test_no_day_week_num
...
-      }]
+    vrp = VRP.lat_lon_two_vehicles
+    vrp[:configuration][:preprocessing] = { partitions: TestHelper.vehicle_and_days_partitions }
+    # ensure one unassigned service :
+    vrp[:services].first[:activity][:timewindows] = [{ end: 100 }]
+    vrp[:vehicles].each{ |v| v[:timewindow] = { start: 200 } }
+    vrp[:configuration][:schedule] = { range_indices: { start: 0, end: 10 }}
+
+    response = post '/0.1/vrp/submit', { api_key: 'demo', vrp: vrp }.to_json, 'CONTENT_TYPE' => 'application/json'
+    result = JSON.parse(response.body)['solutions'].first
```
Comment (braktar) [inline test/api/v01/output_test.rb:83]: The test is dense, could add some comments on what is expected ?

File: test/lib/interpreters/split_clustering_test.rb [hunk 7]
```diff
@@ -744,5 +747,50 @@ def test_clustering_with_sticky_vehicles
       assert_equal 2, (result[:routes].find{ |r| r[:vehicle_id] == 'vehicle_0' }[:activities].collect{ |a| a[:service_id] } & ['service_1', 'service_5']).size
       assert_equal 1, (result[:routes].find{ |r| r[:vehicle_id] == 'vehicle_1' }[:activities].collect{ |a| a[:service_id] } & ['service_12']).size
     end
+
+    def test_list_vehicles
+      # with timewindow
+      vrp = TestHelper.create(VRP.basic)
+      vrp.vehicles.first.timewindow = Models::Timewindow.new(start: 0, end: 10)
+      assert_equal 7, Interpreters::SplitClustering.list_vehicles({ start: 0, end: 6 }, vrp.vehicles).size
+      assert_equal 7, Interpreters::SplitClustering.list_vehicles({ start: 0, end: 10 }, vrp.vehicles).size
```
Comment (braktar) [inline test/lib/interpreters/split_clustering_test.rb:756]: If we have 11 days in the period, why do we only have 7 vehicles ?

Could you add some comments on what is expected and why ?

**Review by braktar — APPROVED** [review]

File: test/lib/interpreters/split_clustering_test.rb (hunk 7 above)
Comment (fonsecadeline) [reply test/lib/interpreters/split_clustering_test.rb:754]: _work_day_ partition provides one cluster per working day (monday, tuesday, ..., sunday), not for every day of schedule. This ensures that all visits of one service will take place at the same week day as other visits. That is why we do need to consider all visits in the same cluster.

I will add some comments in the test.

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Added
+
+- Corresponding vehicle_id is returned for each service's skills if problem is partitioned with vehicle entity [#110](https://github.com/Mapotempo/optimizer-api/pull/110)
```
Comment (fonsecadeline) [inline CHANGELOG.md:7]:
```suggestion
- Corresponding vehicle_id is returned within each service's skills if problem is partitioned with vehicle entity [#110](https://github.com/Mapotempo/optimizer-api/pull/110)
```

**Pull request merged**